경북대 FE 이창목 - week2_step3 [단체 주문 기능 구현하기]

eelkom opened the pull request:
## PR

### ✨ 작업 내용


- 받는 사람 기능 고도화
- [x] [참고 URL](https://kakaotech-mission2-order.pages.dev/)을 살펴보고 받는 사람 기능의 UI를 구현해요.
- [x] 받는 사람을 최대 10명까지 등록 가능하게 해요.
- [x] 10명의 정보가 정확하게 입력되지 않으면 정보가 반영되지 않아요.
- [x] 받는 사람의 전화번호가 중복되면 안돼요.
- [x] 최소 선택 수량은 1개 이상이어야 해요.
- [x] 전화번호는 01012341234 형태로만 받을 수 있어요.
- 주문하기 페이지 고도화
- [x] [참고 URL](https://kakaotech-mission2-order.pages.dev/)을 살펴보고 변경된 기능을 구현해요.
-  (선택) 유효성 검사를 Zod 라이브러리를 사용해서 처리해요.


### 🔍 참고 사항

- 본인만의 기준으로 일관된 코드를 작성해주세요.
- 기능 단위로 나누어 커밋을 해주세요.
- 재사용 가능성이 높은 컴포넌트, Hooks가 보이면 재사용성을 고려하여 만들어요.
- 주문하기 페이지의 폼을 React Hook Form를 사용해서 리팩터링해요. (form 데이터를 useState로 다루는 경우는 없게 해요.)

안녕하세요. 경북대 FE 5팀 이창목 입니다.

eelkom committed:
chore: react-hook-form 설치

eelkom committed:
feat: 받는 사람 기능의 UI를 구현(WIP)

- form 데이터를 useState로다루는 상황
- React Hook Form를 사용해서 리팩터링 예정

Leejin-Yang commented on src/page/OrderPage/components/ReceiverModal.tsx:
```diff
@@ -0,0 +1,194 @@
+  onClick: () => void;
+}
+
+const ReceiverModal = ({ onClick }: ReceiverModalProps) => {
```
모달의 경우 portal을 사용해 부모에 영향을 받지 않도록 루트 밖으로 빼서 작성할 수 있어요. 아래 링크 한 번 참고해보세요 😊

https://react.dev/reference/react-dom/createPortal

eelkom replied:
ReceiverField 컴포넌트 내부에서 모달을 관리하지만 createPortal을 활용해 모달이 body 영역에 그려지도록 수정했습니다!

Leejin-Yang reviewed (changes requested):
창목님 안녕하세요 👋
이번 단계는 라이브러리 학습도 필요해서 시간이 많이 부족했을거라 생각합니다
요구사항 모두 구현하시면 다시 요청 부탁드릴게요 😊

eelkom committed:
refactor: React Hook Form 리팩토링

- form 데이터를 useState로 다루는 경우 없도록 수정
- 받는 사람 최대 10명까지 등록 가능하도록 구현
- 10명의 정보가 정확하게 입력되지 않으면 정보가 반영되지 않도록 유효성
  검사
- 전화번호는 01012341234 형태로만 받을 수 있도록 구현

eelkom committed:
fix: 이름 유효성 검사 오류 수정 & 필요없는 유효성 검사 제거

eelkom committed:
feat: 받는 사람의 전화번호 중복검사 구현

eelkom committed:
refactor: OrderPage 전체 React Hook Form 리팩토링(WIP)

eelkom committed:
refactor: FormProvider로 자식 컴포넌트에게 전달하는 props 제거

- 자식 컴포넌트는 useFormContext 사용

eelkom committed:
refactor: ReceiverInfo 컴포넌트 분리

eelkom committed:
refactor: 사용하지 않는 hook 제거

eelkom committed:
fix: step2 피드백 반영

- 불필요한 주석 및 console.log 제거
- OrderPage form tag 적용
- MyPage 로그아웃 버튼 form tag 제거 후 onClick 적용

eelkom committed:
feat: 주문 성공 시 navigate('/') 구현

eelkom commented:
이번 과제를 진행하며 과제 요구사항을 순차적으로 구현 하다 보니 다음 요구사항을 고려하지 못해서 컴포넌트 구조를 계속 변경해야 했습니다. 최대한 의도에 맞게 리팩토링 하려고 시도는 했지만 리팩토링이 잘 되었는지는 모르겠습니다! 또한 최대한 공식문서를 읽어보고 ai 툴 활용을 줄여보자 라는 생각으로 리팩토링을 시작해 코드가 깔끔하지 못한 것 같습니다. 물론 마지막에는 도움을 많이 받았습니다. 도움이 없었다면 FormProvider 없이 props drilling 으로 구현 했을 것 같습니다.
아래는 제가 이번 과제를 진행하면서 했던 생각들 간단하게 적어봤습니다. 부족한 부분 있다면 알려주시면 감사하겠습니다:)

- 사람을 추가 할 경우(이름,전화번호,수량)와 주문 할 경우(텍스트영역, 이름) 두 번의 유효성 검사가 필요하다고 생각해 useForm 두 개를 생성해 관리하고자 했습니다. 
  - 추후에 useForm의 trigger를 활용해 사람을 추가 할 경우에는 수동? 으로 유효성 검사를 하도록 수정했습니다.
- 주문하기 버튼을 누를 때 유효성 검사가 필요하다고 생각해 OrderPage의 최상위에 useForm을 선언해 사용했습니다.
- 처음에는 FormProvider의 존재를 몰라서 최대한 prop drilling이 발생하지 않는 방법을 고민했습니다. 
- ReceiverField 라는 컴포넌트를 만들어 받는 사람과 관련된 모든 컴포넌트를 관리하고 여기에 useFieldArray를 선언해 사용했습니다.
  - useFieldArray는 ReceiverModal 컴포넌트에서도 필요하고 ReceiverInfoArray 컴포넌트에서도  필요하기 때문에 props로 전달했습니다.
- FormProvider를 최상위 컴포넌트에 감싸고 useFormContext를 활용해 불필요한 props 전달을 줄였습니다.

eelkom committed:
refactor: 컴포넌트 네이밍 변경 및 스타일 수정

Leejin-Yang commented on src/page/Order/components/ReceiverField.tsx:
```diff
@@ -0,0 +1,103 @@
+      <Container>
+        <ButtonArea>
+          <P>받는사람</P>
+          <Button onClick={() => setIsModalOpen(true)}>{!isEmpty ? '수정' : '추가'}</Button>
```
button에 타입을 설정해주지 않아 보내는 사람을 입력했을 때 받는 사람을 추가하려고 누르면 폼이 제출되네요.!

버튼의 기본 타입은 submit이여서 폼 제출과 관련이 없는 일반 버튼이라면 button 타입을 작성해주어야 합니다 😊

Leejin-Yang commented on src/page/Order/components/ReceiverInfo.tsx:
```diff
@@ -0,0 +1,140 @@
+        <h3>받는 사람 {index + 1}</h3>
+        <button type="button" onClick={() => remove(index)}>
+          Delete
+        </button>
```
테스트 중에 삭제 버튼이 잘 보이지 않아 찾기 어려웠어서, 이 부분 ui 개선을 해보면 좋아보여요 😊

Leejin-Yang commented on src/page/Order/components/ReceiverModal.tsx:
```diff
@@ -0,0 +1,200 @@
+          </ReceiverAddedContainer>
+
+          <ButtonArea>
+            <ButtonCancel onClick={onClose}>취소</ButtonCancel>
```
받는 사람 정보를 입력하지 않았을 때, 취소 버튼을 누르면 반영이 되고 있어요.! 요구사항 중에 정확하게 입력되었을 때 반영되어야 한다고 알고 있는데 이 부분 수정이 필요해보여요

Leejin-Yang commented on src/page/Order/components/ReceiverModal.tsx:
```diff
@@ -0,0 +1,200 @@
+              type="button"
+              onClick={() => {
+                append({ name: '', phoneNumber: '', quantity: 1 });
+              }}
```
주문 폼에 바로 등록하고 있어서 바로 반영이 되는 것으로 보이네요

받는 정보는 폼을 나누고 추가하고 등록할 때 최종적으로 유효성 검사를 마쳐서 주문 폼에 등록을 해보면 어떨까요?

Leejin-Yang commented on src/page/Order/index.tsx:
```diff
@@ -0,0 +1,77 @@
+  const onSubmit = () => {
+    alert('주문 성공!');
+    navigate('/');
+  };
```
받는 사람 등록이 되지 않은 경우도 주문을 성공할 수 있네요.!
이 부분도 유효성 검사를 추가해보면 좋아보여요

eelkom replied:
이 부분도 시간이 조금 필요 할 것 같아 다음 단계 진행하면서 반영하겠습니다!

Leejin-Yang commented on src/page/Order/index.tsx:
```diff
@@ -0,0 +1,77 @@
+  });
+
+  const { watch } = orderForm;
+  const watchedReceiverInfos = watch('receiverInfos');
```
폼 관련 로직은 훅으로 분리해보면 어떨까요?

Leejin-Yang commented on src/page/Order/components/ReceiverInfo.tsx:
```diff
@@ -0,0 +1,140 @@
+            <Input
+              type="text"
+              {...register(`receiverInfos.${index}.name`, {
+                required: { value: true, message: '이름을 입력하세요.' },
+                validate: value => (value.trim() === '' ? '이름을 입력해주세요.' : true),
+              })}
```
이런 규칙들도 분리해보면 어떨까요?

eelkom replied:
폼 관련 로직과 규칙은 고민해보고 다음 단계에서 반영하겠습니다!

Leejin-Yang reviewed (changes requested):
마지막 단계까지 잘 달려오셨네요 👍

짧은 시간에 라이브러리를 학습해보고 기존 코드에 반영해보는 것이 많이 어려웠을거라 생각이 듭니다 
그 중에서 FormProvider를 활용해서 props를 줄이려고 한 점 좋아보입니다 😊

다만, 기대한 대로 동작하지 않는 부분이 있어서 핵심적인 부분만 코멘트로 남겨두었는데, 확인 해주시고 다시 요청 보내주세요

eelkom committed:
fix: button type 지정

eelkom committed:
feat: DeleteIcon 추가

eelkom committed:
fix: receiver 정보 즉각 반영 문제 해결

Leejin-Yang commented on src/page/Order/components/ReceiverField.tsx:
```diff
@@ -0,0 +1,109 @@
+  }
+`;
+
+// interface FormInputValues {
+//   receiverInfos: { name: string; phoneNumber: string; quantity: number }[];
+// }
```
사용하지 않는 주석은 제거해주세요 😊

Leejin-Yang commented on src/page/Order/components/ReceiverModal.tsx:
```diff
@@ -0,0 +1,215 @@
+        <ModalContainer>
+          <form
+            onSubmit={receiverInfosForm.handleSubmit(onSubmit)}
+            style={{ display: 'flex', flexDirection: 'column', gap: '16px', height: '100%' }}
```
이 부분도 스타일드로 빼면 좋아보이네요

Leejin-Yang commented on src/page/Order/index.tsx:
```diff
@@ -0,0 +1,80 @@
+  const orderForm = useForm<OrderInfoValues>({
+    defaultValues: { message: '축하해요.', name: '', receiverInfos: [] },
+  });
+  // const orderForm = useForm<OrderInfoValues>({
+  //   defaultValues: { message: '축하해요.', name: '' },
+  // });
```
요기도 주석이 있네요.!

Leejin-Yang reviewed (approved):
이번 미션 진행하느라 고생하셨습니다 👍
다음 단계 시작전에 꼭 폼 관련 내용은 훅으로 분리해보시는걸 추천드려요

이번 미션을 통해 폼 구현과 라이브러리 학습과 활용에 대해 많이 얻어가셨길 바래봅니다.!
리뷰는 여기서 마무리지만, 앞으로도 미션 진행하면서 어려운 점이나 궁금한 점이 있다면 언제든지 DM 남겨주세요. 수고 많았어요 🙌
